File: PYTHON/JunoIBDSignal.py
import numpy as np
from JunoDetector import JunoDetector
from JunoReactorFlux import JunoReactorFlux
import scipy.integrate as integrate

class JunoIBDSignal(object):

    def __init__(self) -> None :

        # binning methods
        self.Ee_edge = []
        self.Ee_edge.append(0.80)
        self.Ee_edge.append(0.94)
        self.Ee_edge.append(7.44)
        self.Ee_edge.append(7.80)
        self.Ee_edge.append(8.20)
        self.Ee_edge.append(12.0)
        self.nbin = []
        self.nbin.append(1)
        self.nbin.append(325)
        self.nbin.append(9)
        self.nbin.append(4)
        self.nbin.append(1)
        self.width = []
        self.width.append( (self.Ee_edge[1] - self.Ee_edge[0]) / self.nbin[0] )
        self.width.append( (self.Ee_edge[2] - self.Ee_edge[1]) / self.nbin[1] )
        self.width.append( (self.Ee_edge[3] - self.Ee_edge[2]) / self.nbin[2] )
        self.width.append( (self.Ee_edge[4] - self.Ee_edge[3]) / self.nbin[3] )
        self.width.append( (self.Ee_edge[5] - self.Ee_edge[4]) / self.nbin[4] )

        self.bin_edge = []
        for i in range(len(self.width)):
            for j in range(self.nbin[i]):
                self.bin_edge.append(self.Ee_edge[i] + self.width[i]*j)
        self.bin_edge.append(self.Ee_edge[-1]) 


        ### Initialization and dataloading...
        self.reactor = JunoReactorFlux()
        self.reactor.CommonInput()

        self.det = JunoDetector()
        self.det.CommonInput()



    def BinnedNe(self, i):

        eff = self.det.GetEfficiency()
        Np  = self.det.GetNproton()

        Ep_low, Ep_high = self.bin_edge[i], self.bin_edge[i+1]
        print("Bin Range [%.5f, %.5f]" %(Ep_low, Ep_high) )
        
        f = lambda Ep, ct, Enu : self.reactor.arrivedFlux(Enu) * self.det.IBDdiffXsec(Enu, ct) * Np * eff * self.det.detector_response(Enu, ct, Ep)

        Ti = integrate.tplquad(f, 1.8, 15, lambda Enu: -1, lambda Enu: 1, lambda Enu, ct: Ep_low, lambda Enu, ct: Ep_high)
        print("Predicted Bin Count = ", Ti)
        return Ti



    def PredPromptSignalSpec(self):
        T = []
        for i in range(len(self.bin_edge)-1) :
            T.append(self.BinnedNe(i))

        print(T)
        return T

File: src/JunoIBDSignal.cc
#include "JunoIBDSignal.hh"

#include <TFile.h>

JunoIBDSignal::JunoIBDSignal(int MO)
{
    m_MO = MO;

    cout << endl;
    cout << "===============================================" << endl;
    cout << "Prediction Mass Ordering is " ;
    if (MO == 1)
        cout << "Normal Ordering !" << endl;
    if (MO == 2)
        cout << "Inverted Ordering !" << endl;
    cout << "===============================================" << endl;
    cout << endl;

    //reactor = new ReactorFlux();
    //reactor->LoadCommonInputs();
    det = new JunoDetector();
    det->LoadCommonInputs();

    JunoConvCore::Initialize(m_MO);
    fVisibleEnergySpectrum = new TF3("fVisibleEnergySpectrum", JunoConvCore::fVisibleSpectrum, 1.8, 15, -1, 1, 0, 12, 1);


    // binning strategy :
    int N = 0;
    for (int i=0; i<5; i++) {
        double bin_width = (zone_edge[i+1] - zone_edge[i]) / nbins[i];
        for (int j=0; j<nbins[i]; j++){
            bin_edge[N] = zone_edge[i] + j*bin_width;
            N++;
        }
    }
    bin_edge[340] = zone_edge[5];

    hPredEvisSpec = new TH1D("hPredEvisSpec", "Predicted visible energy spectrum", 340, bin_edge);

    hWeight = new TH2D("hWeight", "ith reactor ratio in jth bin", 340, 0, 340, 10, 0, 10);

    cout << "\n";
}


JunoIBDSignal::~JunoIBDSignal()
{
    //delete reactor;
    delete det;
    delete fVisibleEnergySpectrum;
}


//double JunoIBDSignal::BinnedNeutrinoEnergySpectrum(double Enu)
//{
//    double arrivedNu = reactor->ArrivedReactorFlux(Enu);
//    double nproton = det->GetNproton();
//    double eff = det->GetEfficiency();
//    double IBDXsec = det->IBDtotXsec(Enu);
//    
//    return arrivedNu * IBDXsec * nproton * eff;
//
//}


void JunoIBDSignal::CalculateReactorBinRatio()
{
    double reactorFlux[10];
    for(int ibin=0; ibin<340; ibin++) {
        double totFlux = 0;
        for( int i=0; i<10; i++) {
            reactorFlux[i] = BinnedVisibleEnergySpectrum(bin_edge[ibin], bin_edge[ibin+1], i);
            totFlux += reactorFlux[i];
        } 
        for( int i=0; i<10; i++) {
            hWeight->SetBinContent(ibin+1, i+1, reactorFlux[i]/totFlux);
        }
    }
}



double JunoIBDSignal::BinnedVisibleEnergySpectrum(double Epmin, double Epmax, int reactorNo)
{
    double sigma = det->Resolution(Epmax);
    double Enumin = Epmin - 5*sigma + 0.8;
    double Enumax = Epmax + 5*sigma + 0.8;
    fVisibleEnergySpectrum->SetParameter(0, reactorNo);
    return fVisibleEnergySpectrum->Integral(Enumin, Enumax, -1, 1, Epmin, Epmax, 1.e-3);

}




TH1D* JunoIBDSignal::PredictedVisibleEnergySpectrum()
{
    CalculateReactorBinRatio();

    for(int ibin=0; ibin<340; ibin++) {
        //cout << ibin << " " << bin_edge[ibin] << " " << bin_edge[ibin+1] << " " << BinnedVisibleEnergySpectrum(bin_edge[ibin], bin_edge[ibin+1]) << endl;
        hPredEvisSpec->SetBinContent(ibin+1, BinnedVisibleEnergySpectrum(bin_edge[ibin], bin_edge[ibin+1], 10) / (bin_edge[ibin+1] - bin_edge[ibin]) * 0.02);
    }
    return hPredEvisSpec;
}



void JunoIBDSignal::Plot()
{
    TFile* ff = new TFile("PredEvisSpec.root", "recreate");
    hPredEvisSpec->Write();
    ff->Close();
}







